### ISSUE TITLE
using MitoZ for annotating mitogenome generated by other assemblers

### ISSUE THREAD

**Issue by orient100** [linzhi2013/MitoZ#4]
Hi,
I've got a mitochondrial assembly file through Novoplasty, and if you continue to use the Mitoz software to annotate and visualize, what can I replace  with  "*" in the input file? I know the '*' in the fasta files indicates that the nucleotide before is a possible deletion/insertion. 
thanks

**Comment by orient100** [comment]
引号里面内容可能不能显示出来，它是中文中的星号。

**Comment by linzhi2013** [comment]
> Hi,
> I've got a mitochondrial assembly file through Novoplasty, and if you continue to use the Mitoz software to annotate and visualize, what can I replace with "_" in the input file? I know the '_' in the fasta files indicates that the nucleotide before is a possible deletion/insertion. 
> thanks


Dear orient100 ,

Thank you very much for using MitoZ and pointing out the problem!

Indeed, we found that mitogenome sequence assembled by NOVOPlasty sometimes contains non-IUPAC (https://www.bioinformatics.org/sms/iupac.html) characters, which will make MitoZ stop during `annotate`, and output an error message like this:

```
Traceback (most recent call last):
  File "/app/release_MitoZ_v2.3/bin/annotate/revise_CDS_pos_v6.py", line 242, in <module>
    pro_seq_start = sub_start_seq.translate(table=table, cds=True, to_stop=False)
  File "/app/anaconda/lib/python3.6/site-packages/Bio/Seq.py", line 1038, in translate
    cds, gap=gap)
  File "/app/anaconda/lib/python3.6/site-packages/Bio/Seq.py", line 2080, in _translate_str
    "First codon '{0}' is not a start codon".format(sequence[:3]))
Bio.Data.CodonTable.TranslationError: First codon 'T*C' is not a start codon

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/anaconda/lib/python3.6/site-packages/Bio/Data/CodonTable.py", line 341, in __getitem__
    self.ambiguous_nucleotide)
  File "/app/anaconda/lib/python3.6/site-packages/Bio/Data/CodonTable.py", line 194, in list_possible_proteins
    x2 = ambiguous_nucleotide_values[c2]
KeyError: '*'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/anaconda/lib/python3.6/site-packages/Bio/Seq.py", line 2107, in _translate_str
    amino_acids.append(forward_table[codon])
  File "/app/anaconda/lib/python3.6/site-packages/Bio/Data/CodonTable.py", line 344, in __getitem__
    raise KeyError(codon)  # stop codon
KeyError: 'T*C'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/release_MitoZ_v2.3/bin/annotate/revise_CDS_pos_v6.py", line 244, in <module>
    pro_seq_start = sub_start_seq.translate(table=table, cds=False, to_stop=False)
  File "/app/anaconda/lib/python3.6/site-packages/Bio/Seq.py", line 1038, in translate
    cds, gap=gap)
  File "/app/anaconda/lib/python3.6/site-packages/Bio/Seq.py", line 2124, in _translate_str
    "Codon '{0}' is invalid".format(codon))
Bio.Data.CodonTable.TranslationError: Codon 'T*C' is invalid
Error occured when running command:
/app/anaconda/bin/python3 /app/release_MitoZ_v2.3/bin/annotate/revise_CDS_pos_v6.py test_mitoscaf.fa test.cds.position.sorted 2 test.cds.position.sorted.revised
```

The crash is caused by `Bio.Data.CodonTable.TranslationError: Codon 'T*C' is invalid`, where `*` is not an IUPAC nucleotide code, which is not allowed by MitoZ (internally, not allowed by BioPython).

*Thus, before using  MitoZ for annotating mitogenome generated by other assemblers, please make sure it does not contain non-IUPAC nucleotide code*. For example, replace them with `N`s. 

In your case, in which the mitogenome was assembled by NOVOPlasty, the `*` has following meanings (see https://github.com/ndierckx/NOVOPlasty for more details):

`A '*' in the fasta output files indicates that the nucleotide before is a possible deletion/insertion. This can occur when the exact length of single nucleotide repeat can't be determined exactly due to systemic Illumina sequencing errors or within repetitive regions. Since this sign can interfere with post processing algorithms it is best resolve them manually or to delete them.`

I would recommend that you resolve or delete the `*` manually  preceding MitoZ. For example, you can map the reads to the mitogenome with the BWA program, then use the `samtools mpileup` command to determine the exact base(s) the `*` represents.  Or simply you do not want to do this, you might just replace the `*` with `N`, the go to MitoZ directly.

Hope the explanation can help you!

Cheers,
Guanliang

**Comment by linzhi2013** [comment]
please feel free to open new question when necessary.

**Comment by LorenaDerezanin** [comment]
Hey there,
I got a similar issue while running ``` annotate ``` command on a mitogenome generated using NOVOplasty. I've made sure the sequence doesn't contain any non-IUPAC codes and that the seqid is short ``` >eira topology=circular ```. It's creating the .tmp dir with intermediary annot. files, properly identifying _Martes martes_ as the most related species, but it doesn't create the  .result dir with the genbank file. I'm using release_MitoZ_v2.4-alpha. I ran it couple of times, tried switching topology  to linear, ran it on the published domestic ferret mitogenome (KT693383.1) but it keeps braking at the same stage:
```
Traceback (most recent call last):
  File "/home/derezanin/my_software/release_MitoZ_v2.4-alpha/bin/annotate/gene_feature.py", line 48, in read_fastaLike2
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/derezanin/my_software/release_MitoZ_v2.4-alpha/bin/annotate/find_contro_region.py", line 364, in <module>
    main()
  File "/home/derezanin/my_software/release_MitoZ_v2.4-alpha/bin/annotate/find_contro_region.py", line 254, in main
    all_seqid_geneCoor = read_featureTable(fea_files=args.fea_files)
  File "/home/derezanin/my_software/release_MitoZ_v2.4-alpha/bin/annotate/find_contro_region.py", line 109, in read_featureTable
    seqid_geneCoor = get_gene_coor(fea_f)
  File "/home/derezanin/my_software/release_MitoZ_v2.4-alpha/bin/annotate/gene_feature.py", line 62, in get_gene_coor
    for rec in records:
RuntimeError: generator raised StopIteration
Error occured when running command:
/home/derezanin/miniconda3/envs/mitozEnv/bin/python3 /home/derezanin/my_software/release_MitoZ_v2.4-alpha/bin/annotate/find_contro_region.py -fa_file /home/derezanin/NO_BACKUP/eira_barbara/mitogenome/mito_assembled/annotated_tayra_mito2/eira_barbara_mitogenome.fasta -PCG_cutoff_file /home/derezanin/my_software/release_MitoZ_v2.4-alpha/bin/profiles/CDS_HMM/Chordata_CDS_length_list -PCG_len_ratio 0.9  -s_rRNA_CM_file /home/derezanin/my_software/release_MitoZ_v2.4-alpha/bin/profiles/rRNA_CM/v1.1_12snew.cm -l_rRNA_CM_file /home/derezanin/my_software/release_MitoZ_v2.4-alpha/bin/profiles/rRNA_CM/v1.1_16snew.cm -rRNA_len_ratio 0.9  -tRNA_num_min 22  -fea_files mitoz_ann5_mitoscaf.fa.cds.ft /data/scratch/derezanin/eira_barbara/mitogenome/MitoZ_run/mitoz_ann5.tmp/mitoz_ann5.annotation/mitoz_ann5_mitoscaf.fa.trna.ft mitoz_ann5_mitoscaf.fa.s-rRNA.ft mitoz_ann5_mitoscaf.fa.l-rRNA.ft -CR_len_min 600  -outfile mitoz_ann5_mitoscaf.fa.control_region.ft
```
What might be the issue here?

Here are the ```mitoscaf.fa.tbl``` from tmp/annotation dirs for both target species as .txt files.
[mtz_df_mitoscaf.fa.txt](https://github.com/linzhi2013/MitoZ/files/4927825/mtz_df_mitoscaf.fa.txt)
[mtz_eira_mitoscaf.fa.txt](https://github.com/linzhi2013/MitoZ/files/4927830/mtz_eira_mitoscaf.fa.txt)

Thank you in advance.

**Comment by linzhi2013** [comment]
Dear LorenaDerezanin,

It seemed to be the Python's problem, see https://stackoverflow.com/questions/51700960/runtimeerror-generator-raised-stopiteration-every-time-i-try-to-run-app

I `annotate` the KT693383.1  with following method

```
$ docker pull guanliangmeng/mitoz:2.3
$ docker run -v $PWD:/project --rm guanliangmeng/mitoz:2.3 /app/release_MitoZ_v2.3/MitoZ.py annotate  --genetic_code 2 --clade Chordata --outprefix test --thread_number 2 --fastafile KT693383.1.fa
```

It ran well and the result is
[test.result.zip](https://github.com/linzhi2013/MitoZ/files/4927996/test.result.zip)


Cheers

**Comment by LorenaDerezanin** [comment]
Thank you for a quick reply, much appreciated :)
So the problem mentioned in the Stack Overflow issue might be also valid for the Python v.3.8. in the conda env I've created for MitoZ v.2.4. 
So after conda cleanup, I created the new conda env for MitoZ v.2.3. release from yaml file with list of pkgs mentioned in issue #47 including python 3.6., and ran ```annotate``` on domestic ferret (KT693383.1) genome to recreate your results, and once again on my target species. Everything worked out well. 

Thank you once again :)
